// node-simconnect/dist/enums/SimConnectPeriod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimConnectPeriod = void 0;
var SimConnectPeriod;
(function (SimConnectPeriod) {
    /** Specifies that the data is not to be sent. */
    SimConnectPeriod[SimConnectPeriod["NEVER"] = 0] = "NEVER";
    /** Specifies that the data should be sent once only. Note that this is not an efficient way of receiving
     * data frequently, use one of the other periods if there is a regular frequency to the data request. */
    SimConnectPeriod[SimConnectPeriod["ONCE"] = 1] = "ONCE";
    /** Specifies that the data should be sent every visual (rendered) frame. */
    SimConnectPeriod[SimConnectPeriod["VISUAL_FRAME"] = 2] = "VISUAL_FRAME";
    /** Specifies that the data should be sent every simulated frame, whether that frame is rendered or not. */
    SimConnectPeriod[SimConnectPeriod["SIM_FRAME"] = 3] = "SIM_FRAME";
    /** Specifies that the data should be sent once every second. */
    SimConnectPeriod[SimConnectPeriod["SECOND"] = 4] = "SECOND";
})(SimConnectPeriod = exports.SimConnectPeriod || (exports.SimConnectPeriod = {}));
module.exports = {
    SimConnectPeriod,
};
