// node-simconnect/dist/enums/SimConnectDataType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimConnectDataType = void 0;
var SimConnectDataType;
(function (SimConnectDataType) {
    /** Invalid data type */
    SimConnectDataType[SimConnectDataType["INVALID"] = 0] = "INVALID";
    /** 32-bit integer number */
    SimConnectDataType[SimConnectDataType["INT32"] = 1] = "INT32";
    /** 64-bit integer number */
    SimConnectDataType[SimConnectDataType["INT64"] = 2] = "INT64";
    /** 32-bit floating-point number (float) */
    SimConnectDataType[SimConnectDataType["FLOAT32"] = 3] = "FLOAT32";
    /** 64-bit floating-point number (double) */
    SimConnectDataType[SimConnectDataType["FLOAT64"] = 4] = "FLOAT64";
    /** Fixed-length string, 8-byte */
    SimConnectDataType[SimConnectDataType["STRING8"] = 5] = "STRING8";
    /** Fixed-length string, 32-byte */
    SimConnectDataType[SimConnectDataType["STRING32"] = 6] = "STRING32";
    /** Fixed-length string, 64-byte */
    SimConnectDataType[SimConnectDataType["STRING64"] = 7] = "STRING64";
    /** Fixed-length string, 128-byte */
    SimConnectDataType[SimConnectDataType["STRING128"] = 8] = "STRING128";
    /** Fixed-length string, 256-byte */
    SimConnectDataType[SimConnectDataType["STRING256"] = 9] = "STRING256";
    /** Fixed-length string, 260-byte */
    SimConnectDataType[SimConnectDataType["STRING260"] = 10] = "STRING260";
    /** Variable-length string */
    SimConnectDataType[SimConnectDataType["STRINGV"] = 11] = "STRINGV";
    /** {@link InitPosition} data structure */
    SimConnectDataType[SimConnectDataType["INITPOSITION"] = 12] = "INITPOSITION";
    /** {@link MarkerState} data structure */
    SimConnectDataType[SimConnectDataType["MARKERSTATE"] = 13] = "MARKERSTATE";
    /** {@link Waypoint} data structure */
    SimConnectDataType[SimConnectDataType["WAYPOINT"] = 14] = "WAYPOINT";
    /** {@link LatLonAlt} data structure */
    SimConnectDataType[SimConnectDataType["LATLONALT"] = 15] = "LATLONALT";
    /** {@link XYZ} data structure */
    SimConnectDataType[SimConnectDataType["XYZ"] = 16] = "XYZ";
    /** enum limit */
    SimConnectDataType[SimConnectDataType["MAX"] = 17] = "MAX";
})(SimConnectDataType = exports.SimConnectDataType || (exports.SimConnectDataType = {}));
module.exports = {
    SimConnectDataType,
};
