// node-simconnect/dist/enums/TextResult.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextResult = void 0;
var TextResult;
(function (TextResult) {
    /**   Specifies that the user has selected the menu item. */
    TextResult[TextResult["MENU_SELECT_1"] = 0] = "MENU_SELECT_1";
    /**   Specifies that the user has selected the menu item. */
    TextResult[TextResult["MENU_SELECT_2"] = 1] = "MENU_SELECT_2";
    /**   Specifies that the user has selected the menu item. */
    TextResult[TextResult["MENU_SELECT_3"] = 2] = "MENU_SELECT_3";
    /**   Specifies that the user has selected the menu item. */
    TextResult[TextResult["MENU_SELECT_4"] = 3] = "MENU_SELECT_4";
    /**   Specifies that the user has selected the menu item. */
    TextResult[TextResult["MENU_SELECT_5"] = 4] = "MENU_SELECT_5";
    /**   Specifies that the user has selected the menu item. */
    TextResult[TextResult["MENU_SELECT_6"] = 5] = "MENU_SELECT_6";
    /**   Specifies that the user has selected the menu item. */
    TextResult[TextResult["MENU_SELECT_7"] = 6] = "MENU_SELECT_7";
    /**   Specifies that the user has selected the menu item. */
    TextResult[TextResult["MENU_SELECT_8"] = 7] = "MENU_SELECT_8";
    /**   Specifies that the user has selected the menu item. */
    TextResult[TextResult["MENU_SELECT_9"] = 8] = "MENU_SELECT_9";
    /**   Specifies that the user has selected the menu item. */
    TextResult[TextResult["MENU_SELECT_10"] = 9] = "MENU_SELECT_10";
    /**   Specifies that the menu or text identified by the EventID is now on display. */
    TextResult[TextResult["DISPLAYED"] = 65536] = "DISPLAYED";
    /**   Specifies that the menu or text identified by the EventID is waiting in a queue. */
    TextResult[TextResult["QUEUED"] = 65537] = "QUEUED";
    /**   Specifies that the menu or text identified by the EventID has been removed from the queue. */
    TextResult[TextResult["REMOVED"] = 65538] = "REMOVED";
    /**    Specifies that the menu or text identified by the EventID has been replaced in the queue. */
    TextResult[TextResult["REPLACED"] = 65539] = "REPLACED";
    /**   Specifies that the menu or text identified by the EventID has timed-out and is no longer on display. */
    TextResult[TextResult["TIMEOUT"] = 65540] = "TIMEOUT";
})(TextResult = exports.TextResult || (exports.TextResult = {}));
module.exports = {
    TextResult,
};
