// node-simconnect/dist/dto/InitPosition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Airspeed = exports.InitPosition = void 0;
var Airspeed;
(function (Airspeed) {
    Airspeed[Airspeed["Keep"] = -2] = "Keep";
    Airspeed[Airspeed["Cruise"] = -1] = "Cruise";
})(Airspeed || (Airspeed = {}));
exports.Airspeed = Airspeed;
class InitPosition {
    constructor() {
        this.latitude = 0;
        this.longitude = 0;
        this.altitude = 0;
        this.pitch = 0;
        this.bank = 0;
        this.heading = 0;
        this.onGround = false;
        this.airspeed = 0;
    }
    read(buffer) {
        this.latitude = buffer.readDouble();
        this.longitude = buffer.readDouble();
        this.altitude = buffer.readDouble();
        this.pitch = buffer.readDouble();
        this.bank = buffer.readDouble();
        this.heading = buffer.readDouble();
        this.onGround = buffer.readInt() !== 0;
        this.airspeed = buffer.readInt();
    }
    write(buffer) {
        buffer.writeDouble(this.latitude);
        buffer.writeDouble(this.longitude);
        buffer.writeDouble(this.altitude);
        buffer.writeDouble(this.pitch);
        buffer.writeDouble(this.bank);
        buffer.writeDouble(this.heading);
        buffer.writeInt(this.onGround ? 1 : 0);
        buffer.writeInt(this.airspeed);
    }
}
exports.InitPosition = InitPosition;
