// node-simconnect/dist/facility/FacilityVOR.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FacilityVOR = void 0;
const FacilityNDB_1 = require("./FacilityNDB");
class FacilityVOR extends FacilityNDB_1.FacilityNDB {
    constructor(data) {
        super(data);
        this.flags = data.readInt();
        this.localizer = data.readFloat();
        this.glideLat = data.readDouble();
        this.glideLon = data.readDouble();
        this.glideAlt = data.readDouble();
        this.glideSlipeAngle = data.readFloat();
    }
    hasFlag(flag) {
        return (this.flags & flag) !== 0;
    }
    hasNavSignal() {
        return this.hasFlag(FacilityVOR.HAS_NAV_SIGNAL);
    }
    hasLocalizer() {
        return this.hasFlag(FacilityVOR.HAS_LOCALIZER);
    }
    hasGlideSlope() {
        return this.hasFlag(FacilityVOR.HAS_GLIDE_SLOPE);
    }
    hasDME() {
        return this.hasFlag(FacilityVOR.HAS_DME);
    }
}
exports.FacilityVOR = FacilityVOR;
FacilityVOR.HAS_NAV_SIGNAL = 0x00000001;
FacilityVOR.HAS_LOCALIZER = 0x00000002;
FacilityVOR.HAS_GLIDE_SLOPE = 0x00000004;
FacilityVOR.HAS_DME = 0x00000008;
