// node-simconnect/dist/enums/ClientDataPeriod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientDataPeriod = void 0;
var ClientDataPeriod;
(function (ClientDataPeriod) {
    /** Specifies that the data is not to be sent. */
    ClientDataPeriod[ClientDataPeriod["NEVER"] = 0] = "NEVER";
    /** Specifies that the data should be sent once only. Note that this is not an efficient way of receiving
     * data frequently, use one of the other periods if there is a regular frequency to the data request. */
    ClientDataPeriod[ClientDataPeriod["ONCE"] = 1] = "ONCE";
    /** Specifies that the data should be sent every visual (rendered) frame. */
    ClientDataPeriod[ClientDataPeriod["VISUAL_FRAME"] = 2] = "VISUAL_FRAME";
    /** Specifies that the data should be sent whenever it is set. */
    ClientDataPeriod[ClientDataPeriod["ON_SET"] = 3] = "ON_SET";
    /** Specifies that the data should be sent once every second. */
    ClientDataPeriod[ClientDataPeriod["SECOND"] = 4] = "SECOND";
})(ClientDataPeriod = exports.ClientDataPeriod || (exports.ClientDataPeriod = {}));
