// node-simconnect/dist/enums/TextType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextType = void 0;
var TextType;
(function (TextType) {
    /** specify scrolling text in the named color */
    TextType[TextType["SCROLL_BLACK"] = 0] = "SCROLL_BLACK";
    /** specify scrolling text in the named color */
    TextType[TextType["SCROLL_WHITE"] = 1] = "SCROLL_WHITE";
    /** specify scrolling text in the named color */
    TextType[TextType["SCROLL_RED"] = 2] = "SCROLL_RED";
    /** specify scrolling text in the named color */
    TextType[TextType["SCROLL_GREEN"] = 3] = "SCROLL_GREEN";
    /** specify scrolling text in the named color */
    TextType[TextType["SCROLL_BLUE"] = 4] = "SCROLL_BLUE";
    /** specify scrolling text in the named color */
    TextType[TextType["SCROLL_YELLOW"] = 5] = "SCROLL_YELLOW";
    /** specify scrolling text in the named color */
    TextType[TextType["SCROLL_MAGENTA"] = 6] = "SCROLL_MAGENTA";
    /** specify scrolling text in the named color */
    TextType[TextType["SCROLL_CYAN"] = 7] = "SCROLL_CYAN";
    /** specify static text in the named color */
    TextType[TextType["PRINT_BLACK"] = 256] = "PRINT_BLACK";
    /** specify static text in the named color */
    TextType[TextType["PRINT_WHITE"] = 257] = "PRINT_WHITE";
    /** specify static text in the named color */
    TextType[TextType["PRINT_RED"] = 258] = "PRINT_RED";
    /** specify static text in the named color */
    TextType[TextType["PRINT_GREEN"] = 259] = "PRINT_GREEN";
    /** specify static text in the named color */
    TextType[TextType["PRINT_BLUE"] = 260] = "PRINT_BLUE";
    /** specify static text in the named color */
    TextType[TextType["PRINT_YELLOW"] = 261] = "PRINT_YELLOW";
    /** specify static text in the named color */
    TextType[TextType["PRINT_MAGENTA"] = 262] = "PRINT_MAGENTA";
    /** specify static text in the named color */
    TextType[TextType["PRINT_CYAN"] = 263] = "PRINT_CYAN";
    /** specify that the text is for a menu */
    TextType[TextType["MENU"] = 512] = "MENU";
})(TextType = exports.TextType || (exports.TextType = {}));
